server/core/src/main/java/org/infinispan/server/core/transport/NonRecursiveEventLoopGroup.java

```diff
@@ -0,0 +1,146 @@
+         throw new IllegalArgumentException("Provided multi threaded event loop group must have at least 2 executors, only has " + executors);
+      }
+      this.eventLoopGroup = eventLoopGroup;
+      // Try to get an event loop a bit more than the number of executors
+      this.loopAttempts = executors << 1;
+   }
+
+   @Override
+   protected EventLoopGroup delegate() {
+      return eventLoopGroup;
+   }
+
+   @Override
+   public void execute(Runnable command) {
+      getExecutorNotInEventLoop().execute(command);
+   }
+
+   @Override
+   public Future<?> submit(Runnable task) {
+      return getExecutorNotInEventLoop().submit(task);
+   }
+
+   @Override
+   public <T> Future<T> submit(Callable<T> task) {
+      return getExecutorNotInEventLoop().submit(task);
+   }
+
+   @Override
+   public <T> Future<T> submit(Runnable task, T result) {
+      return getExecutorNotInEventLoop().submit(task, result);
+   }
+
+   /**
+    * This reserves the current thread's executor so that no other executor may submit tasks to it including itself.
+    * This should be used if the calling code will end up blocking when it shouldn't and is a sign of code that
+    * should eventually be refactored but provides a way to share the netty event loop
+    */
+   public static void reserveCurrentThread() {
+      EventExecutor thisExecutor = ThreadExecutorMap.currentExecutor();
+      if (thisExecutor != null) {
+         boolean reserved = reserveExecutors.add(thisExecutor);
+         assert reserved;
+      }
+   }
+
+   /**
+    * No longer reserves the current thread and allows it to be submitted to. This should always be invoked when
+    * the blocking in the invoking thread is complete as soon as possible.
+    */
+   public static void unreserveCurrentThread() {
+      EventExecutor thisExecutor = ThreadExecutorMap.currentExecutor();
+      if (thisExecutor != null) {
+         boolean unreserved = reserveExecutors.remove(thisExecutor);
+         assert unreserved;
+      }
+   }
+
+   private EventExecutor getExecutorNotInEventLoop() {
+      for (int i = 0; i < loopAttempts; ++i) {
+         EventExecutor eventExecutor = eventLoopGroup.next();
+         // We don't want to submit a task to our current thread as some callers may block waiting for it to complete.
+         if (eventExecutor.inEventLoop()) {
+            log.tracef("Skipped submitting task to %s as it is the current event loop - trying another", eventExecutor);
+            continue;
+         }
+         // We also don't want to submit to a reserved executor as it is currently blocking waiting for another caller
+         if (reserveExecutors.contains(eventExecutor)) {
+            log.tracef("Skipped submitting task to %s as it is currently reserved - trying another", eventExecutor);
+            continue;
+         }
+
+         return eventExecutor;
+      }
+      log.fatalf("No executor was found - replaying to trace issue - current reserveExecutors are %s", reserveExecutors);
```

Comment:
I think the `fatal` messages have to be defined in the `Log` interface, right?

Reply:
I am torn, because this is will have been caused by a programming bug and the user can't really do anything about it. I can change them to DEBUG I guess.

Reply:
the question is like: how does it help you find "a bug"?

Reply:
If you are asking why have the trace messages at all and just throw the exeception then the answer is it tells you why all the various executors were not able to be chosen, either by being the same event loop or it was reserved.

We could just remove the log messages I guess. This shouldn't really happen now, it was very helpful while finding the various areas. But since it should only show up for new code, it is probably fine.